README
grader-cli version 2.2

CONFIGURATION
--------------

There are a couple of things you can configure manually in the script. (unfortunately not much yet..)

Those are:
	* input/output filename format (defaults are 'input.#.txt' and	'output.#.txt')
	* time limit (default is 1 second)
	* the compiler and compiling options for each language (the language is detected based on the file extension)

Note: the '#' character in the input/output filename description will be replaced by 1, 2, 3, ... for each test case. For example 'input.#.txt' means that the input files are: input.1.txt , input.2.txt , ... up to the maximum numbered test case existing in the current directory.

USAGE
------

Just run the script with one parameter, the source code filename. 

Example:
	./grader.sh test.cpp

	The above command will first compile test.cpp and then run the executable for each test case of the form described by the user.
	After each run (and if the program terminates with exit code 0 within the time limit), the output is judged based on the given correct output files.

*** IMPORTANT NOTES ***
-----------------------
	* Your program should read data from stdin and write on stdout. The required input/output redirections are handled automatically by the script.
	* Output is judged based on the given correct output files using the diff tool. As a result, the grading script might give wrong results for problems in which more than one outputs are considered correct by the problem statement.
	* The script, the source code and the test cases should be placed under the same directory.


CHANGELOG
Bug fixes [Feb. 22, 2020]:

1. Fixed escape sequences for colors. It seems that \e is not supported in some systems so I replaced it with \033.
2. Time Limit Exceeded used to be interpreted as Runtime error in some system. Temporary fix applied: Interpret both return codes 152 and 137 as "terminated by SIGXCPU" which is the signal ulimit is supposed to send when CPU time is exceeded. A more permanent/universal solution is still pending.


grader-cli-2.2.sh
#!/bin/bash
## Version 2.2

## ****** BEGIN - Configuration *******

# Input/Output format
# '#' will be replaced by 1, 2, ...
IN="input.#.txt"
OUT="output.#.txt"

# Limits
TL=1 # Time limit (in seconds)

# Compiling options
CPP="g++ -O2 -w -lm" # C++
C="gcc -O2 -w -lm" # C
PAS="gpc" # Pascal

## ****** END - Configuration ********

#--------------------------------------------------

# Color Codes
red='\e[0;31m'
RED='\e[1;31m'
GREEN='\e[1;32m'
green='\e[0;32m'
blue='\e[0;34m'
BLUE='\e[1;34m'
cyan='\e[0;36m'
CYAN='\e[1;36m'
NC='\e[0m' # No Color

# Cleanup on exit
rm -f .overview .compiler_report .time_info .$1.out
trap "{ rm -f .overview .compiler_report .time_info .$1.out; }" SIGINT SIGTERM EXIT

if [ $# -ne 1 ]
then
	echo "Usage: $0 source_code"
	echo "   e.g. $0 test.cpp"
	echo "   use the above to grade file test.cpp"
	exit 2
fi

# Language detection
LANG=`echo $1 | awk -F . '{print $NF}'`
if [ "$LANG" == "cpp" ]
then
	COMPILER="$CPP $1 2> .compiler_report" # C++
elif [ "$LANG" == "c" ]
then
	COMPILER="$C $1 2> .compiler_report" # C
elif [ "$LANG" == "pas" ]
then
	COMPILER="$PAS $1 2> .compiler_report" # Pascal
fi

# Compilation
echo -e " ${CYAN}* Compiling source code${NC}";
echo "$COMPILER" | sh
if [ $? -ne 0 ]
then
	echo -e " ${RED}X Compilation Error${NC}";
	cat .compiler_report;
	exit 1;
fi

echo -e " ${GREEN}* Successful compilation!${NC}";
echo

ulimit -t $TL;

rm -rf .overview;
CORRECT=0
MAX_N=50

for (( i=1; i<=$MAX_N; i++))
do
	TEST_CASE_IN=`echo $IN | sed "s/#/$i/g"`
	TEST_CASE_OUT=`echo $OUT | sed "s/#/$i/g"`

	# If i-th test case doesn't exist then stop here.
	if [ ! -e $TEST_CASE_IN ]
	then
		break
	fi
	echo -e "${BLUE}Test case $i:${NC}";

	time -p (./a.out < $TEST_CASE_IN > .$1.out) 2> .time_info;

	EX_CODE=$?;
	if [ $EX_CODE -eq 137 ]
	then
		echo -e " ${RED}X TLE: Time Limit Exceeded${NC}";
		echo -n "T" >> .overview;
	elif [ $EX_CODE -ne 0 ]
	then
		echo -e " ${RED}X RE: Runtime Error${NC}";
		echo -n "E" >> .overview;
	else
		PROG_TIME=`cat .time_info | grep real | cut -d" " -f2`;
		diff --strip-trailing-cr .$1.out $TEST_CASE_OUT > /dev/null
		if [ $? -eq 0 ]
		then
			echo -e " ${GREEN}* OK${NC} [$PROG_TIME]"
			echo -n "*" >> .overview
			CORRECT=`expr $CORRECT + 1`
		else
			echo -e " ${RED}X WA: Wrong Answer${NC} [$PROG_TIME]"
			echo -n "X" >> .overview
		fi
	fi

	echo;
done
N=`expr $i - 1`

echo;

echo >> .overview;
echo -n "Overview: "; cat .overview
if [ $CORRECT -ne $N ]
then
	echo -e "${RED}X${NC}: Wrong Answer, ${RED}T${NC}: Time Limit Exceeded, ${RED}E${NC}: Probably runtime error"
	echo
fi

echo -n "$CORRECT / $N"
if [ $CORRECT -eq $N ]
then
	echo -en "   ${GREEN}AWWW YEAH :D${NC}"
fi
echo


grader.sh
#!/bin/bash
## Version 2.4

## ****** BEGIN - Configuration *******

# Input/Output format
# '#' will be replaced by 1, 2, ...
IN="input.#.txt"
OUT="output.#.txt"

# Limits
TL=1 # Time limit (in seconds)

# Compiling options
CPP="g++ -O2 -w -lm" # C++
C="gcc -O2 -w -lm" # C
PAS="gpc" # Pascal

## ****** END - Configuration ********

#--------------------------------------------------

#Copyright (c) 2012 Makis Arsenis

#Permission is hereby granted, free of charge, to any person obtaining a copy
#of this software and associated documentation files (the "Software"), to
#deal in the Software without restriction, including without limitation the
#rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
#sell copies of the Software, and to permit persons to whom the Software is
#furnished to do so, subject to the following conditions:

#The above copyright notice and this permission notice shall be included in
#all copies or substantial portions of the Software.

#THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
#FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
#IN THE SOFTWARE.

# Color Codes
red='\033[0;31m'
RED='\033[1;31m'
GREEN='\033[1;32m'
green='\033[0;32m'
blue='\033[0;34m'
BLUE='\033[1;34m'
cyan='\033[0;36m'
CYAN='\033[1;36m'
NC='\033[0m' # No Color

# Cleanup on exit
rm -f .overview .compiler_report .time_info .$1.out
trap "{ rm -f .overview .compiler_report .time_info .$1.out; }" SIGINT SIGTERM EXIT

if [ $# -ne 1 ]
then
  echo "Usage: $0 source_code"
  echo "   e.g. $0 test.cpp"
  echo "   use the above to grade file test.cpp"
  exit 2
fi

# Language detection
LANG=`echo $1 | awk -F . '{print $NF}'`
if [ "$LANG" == "cpp" ]
then
  COMPILER="$CPP $1 2> .compiler_report" # C++
elif [ "$LANG" == "c" ]
then
  COMPILER="$C $1 2> .compiler_report" # C
elif [ "$LANG" == "pas" ]
then
  COMPILER="$PAS $1 2> .compiler_report" # Pascal
fi

# Compilation
echo -e " ${CYAN}* Compiling source code${NC}";
echo "$COMPILER" | sh
if [ $? -ne 0 ]
then
  echo -e " ${RED}X Compilation Error${NC}";
  cat .compiler_report;
  exit 1;
fi

echo -e " ${GREEN}* Successful compilation!${NC}";
echo

ulimit -t $TL;

rm -rf .overview;
CORRECT=0
MAX_N=50

for (( i=1; i<=$MAX_N; i++))
do
  TEST_CASE_IN=`echo $IN | sed "s/#/$i/g"`
  TEST_CASE_OUT=`echo $OUT | sed "s/#/$i/g"`

  # If i-th test case doesn't exist then stop here.
  if [ ! -e $TEST_CASE_IN ]
  then
    break
  fi
  echo -e "${BLUE}Test case $i:${NC}";

  time -p (./a.out < $TEST_CASE_IN > .$1.out) 2> .time_info;

  EX_CODE=$?;
  if [ $EX_CODE -eq 137 ] || [ $EX_CODE -eq 152 ]
  then
    echo -e " ${RED}X TLE: Time Limit Exceeded${NC}";
    echo -n "T" >> .overview;
  elif [ $EX_CODE -ne 0 ]
  then
    echo -e " ${RED}X RE: Runtime Error${NC}";
    echo -n "E" >> .overview;
  else
    PROG_TIME=`cat .time_info | grep real | cut -d" " -f2`;
    diff --strip-trailing-cr .$1.out $TEST_CASE_OUT > /dev/null
    if [ $? -eq 0 ]
    then
      echo -e " ${GREEN}* OK${NC} [$PROG_TIME]"
      echo -n "*" >> .overview
      CORRECT=`expr $CORRECT + 1`
    else
      echo -e " ${RED}X WA: Wrong Answer${NC} [$PROG_TIME]"
      echo -n "X" >> .overview
    fi
  fi

  echo;
done
N=`expr $i - 1`

echo;

echo >> .overview;
echo -n "Overview: "; cat .overview
if [ $CORRECT -ne $N ]
then
  echo -e "${RED}X${NC}: Wrong Answer, ${RED}T${NC}: Time Limit Exceeded, ${RED}E${NC}: Probably runtime error"
  echo
fi

echo -n "$CORRECT / $N"
if [ $CORRECT -eq $N ]
then
  echo -en "   ${GREEN}All test cases passed!!${NC}"
fi
echo
